### PR TITLE
refactor: make fee storage more efficient

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

The following PR updates the ics29 proto models to accommodate for more efficient use of storage. 
This includes the introduction of a new `PacketFee` type which *excludes* the packet identifier (channelID/portID/sequence). 
Identified Fees will be maintained for users and genesis state ux. 

This PR uses the proto definitions proposed in #915. 

Follow-up PRs will address:
- Updating of GenesisState #957 
- Updating of existing queries for `IncentivizedPacket` & `IncentivizedPackets` #967 

closes: #760

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by damiannolan** [comment]
There is a small dip in coverage but should be pretty trivial to resolve in following PRs. Didn't want to over bloat these unnecessarily. I think its just from the addition of `NewFee()`, `NewPacketFee()`...etc in `types`

File: modules/apps/29-fee/keeper/msg_server.go
Comment (damiannolan) [inline modules/apps/29-fee/keeper/msg_server.go:60]: This todo will be addressed in a follow up. I think the `MsgPayPacketFeeAsync` should look something like this:

```
message MsgPayPacketFeeAsync {
    ibc.core.channel.v1.PacketId packet_id = 1;
    PacketFee packet_fee = 2;
}
```

**Review by seantking — APPROVED** [review]
LGTM but can we merge this after my currently open PRs are merged? Just to limit the merge conflicts. I'll try to get everything merged today.

**Comment by damiannolan** [comment]
> LGTM but can we merge this after my currently open PRs are merged? Just to limit the merge conflicts. I'll try to get everything merged today.

Sounds like a plan to me! I was planning that I'd stack up these PRs and get them in together. Happy to get the others merged first 👍

File: modules/apps/29-fee/keeper/keeper.go
Comment (AdityaSripal) [inline modules/apps/29-fee/keeper/keeper.go:286]: Why was this function renamed?

File: modules/apps/29-fee/keeper/keeper.go
Comment (AdityaSripal) [inline modules/apps/29-fee/keeper/keeper.go:299]: I don't understand the changes you've made to these iterate functions

**Review by AdityaSripal — COMMENTED** [review]
LGTM, just confused by iterate fns

File: proto/ibc/applications/fee/v1/fee.proto
Comment (colin-axner) [inline proto/ibc/applications/fee/v1/fee.proto:61]: should this be `PacketFees`?

File: modules/apps/29-fee/keeper/keeper.go
Comment (damiannolan) [reply modules/apps/29-fee/keeper/keeper.go:286]: Aha, yes. Sorry for the confusion here. This stems from me adding [IterateIdentifiedChannelFeesInEscrow](IterateIdentifiedChannelFeesInEscrow) in #915. This was added to not break existing code and reduce the PR diff, but leads to confusion as there shouldn't be two iteration functions.

In this PR I've changed `IterateChannelFeesInEscrow` to `IteratePacketFeesInEscrow` - this difference being that this function now iterates `PacketFees` (multiple fees without packet ID) instead of `IdentifiedPacketFee` (singular with packet ID). I updated the naming as I thought it suited a bit better. 
The other iteration function can be removed after #957 and #967. 

Sorry about the weird diff here.

File: proto/ibc/applications/fee/v1/fee.proto
Comment (damiannolan) [reply proto/ibc/applications/fee/v1/fee.proto:61]: Good question. It could be `PacketFees` (non repeated), but `PacketFees` is just really a wrapper type for marshalling the fees as bytes to store in state. 
I figured using regular `repeated PacketFee` here just reduced having to reference an additional child if that makes sense. e.g. instead of `identifiedFees.PacketFees.PacketFees` we can just use `identifiedFees.PacketFees`

Comment (damiannolan) [reply proto/ibc/applications/fee/v1/fee.proto:61]: I like this idea. I can take a look later today 👍

File: modules/apps/29-fee/keeper/keeper.go
Comment (colin-axner) [reply modules/apps/29-fee/keeper/keeper.go:286]: Do we need identified packet iteration for `Add gRPC/cli for incentivized packets on a certain channel` in #871?

For `IteratePacketFeesInEscrow` this is for a specific channel? So maybe `IteratePacketFeesInEscrowForChannel` or `IteratePacketFeesInEscrowOnChannel`?

File: proto/ibc/applications/fee/v1/fee.proto
Comment (damiannolan) [reply proto/ibc/applications/fee/v1/fee.proto:61]: I don't think this is possible. If I understood the suggestion correctly it would be to do away with `PacketFees` and have that defined as an alias in the Go code and use `(gogoproto.castrepeated)`:
```
type PacketFees []PacketFee
```

I believe we still need the proto defined type `PacketFees` as the type we store [here](https://github.com/cosmos/ibc-go/blob/damian/760-effecient-storage/modules/apps/29-fee/keeper/keeper.go#L361) must implement `codec.ProtoMarshaller` and using an alias type will not have the proto specific methods needed to marshal bytes.

Comment (colin-axner) [reply proto/ibc/applications/fee/v1/fee.proto:61]: Yea that seems correct. Looks like the SDK always does some workaround to directly setting a slice of a type (splitting up the type into a key/value or adding some other value to the key to create mapping to each element in the slice)

I'm in favor of the current implementation :+1:

File: modules/apps/29-fee/keeper/msg_server.go
Comment (damiannolan) [reply modules/apps/29-fee/keeper/msg_server.go:60]: Opened #975

File: modules/apps/29-fee/keeper/keeper.go
Comment (colin-axner) [reply modules/apps/29-fee/keeper/keeper.go:287]: Yea it'd need pagination. Good point

The function name sounds fine to me. I guess if the function iterated all packet fees it'd be titled `IterateAllPacketFees`

random sidenote, but thoughts on making our iteration functions private? I feel like only functions with `Get` should be public and if it starts with `Iterate` it should be private

File: modules/apps/29-fee/types/fee.go
Comment (colin-axner) [inline modules/apps/29-fee/types/fee.go:11]: what happened to this constructor?

Comment (damiannolan) [reply modules/apps/29-fee/types/fee.go:11]: Replaced by `NewPacketFees()` on line 20 below. 

The `IdentifiedPacketFees` struct used for Genesis is a repeated type so the `[]IdentifiedPacketFees` is returned directly from `GetAllIdentifiedPacketFees` and used in `ExportGenesis`. See #957 -> here https://github.com/cosmos/ibc-go/pull/957/files#diff-b7d50f2bf0988eeab4d6734bd38c10ebf00bda7451e4a6730575de154406937cR330

File: modules/apps/29-fee/keeper/keeper.go
Comment (damiannolan) [reply modules/apps/29-fee/keeper/keeper.go:287]: Yeah I agree, the less functions exposed the better I guess! I'd be happy to make the iterate functions private. We should do that in a follow up if so tho!

**Review by AdityaSripal — APPROVED** [review]

**Pull request merged**
